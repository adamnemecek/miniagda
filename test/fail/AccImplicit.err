MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "AccImplicit.ma" ---
--- scope checking ---
--- type checking ---
type  Acc : ^(A : Set) -> ^(Lt : A -> A -> Set) -> (b : A) -> Set
term  acc : .[A : Set] -> .[Lt : A -> A -> Set] -> .[b : A] -> ^(accOut : (a : A) -> Lt a b -> Acc A Lt a) -> < acc accOut : Acc A Lt b >
term  accOut : .[A : Set] -> .[Lt : A -> A -> Set] -> (b : A) -> (acc : Acc A Lt b) -> (a : A) -> Lt a b -> Acc A Lt a
{ accOut [A] [Lt] b (acc #accOut) = #accOut
}
type  Nat : Set
term  zero : < zero : Nat >
term  succ : ^(y0 : Nat) -> < succ y0 : Nat >
type  R : ^ Nat -> ^ Nat -> Set
term  r1 : .[x : Nat] -> < r1 x : R (succ (succ x)) (succ zero) >
term  r2 : < r2 : R (succ zero) zero >
term  acc2 : (n : Nat) -> Acc Nat R (succ (succ n))
term  acc2 = \ n -> acc (\ a -> \ p -> case p : R a (succ (succ n))
                          {})
term  aux1 : (a : Nat) -> (p : R a (succ zero)) -> Acc Nat R a
{ aux1 (succ (succ x)) (r1 [.x]) = acc2 x
}
term  acc1 : Acc Nat R (succ zero)
term  acc1 = acc aux1
term  aux0 : (a : Nat) -> (p : R a zero) -> Acc Nat R a
{ aux0 .(succ zero) (r2) = acc1
}
term  acc0 : Acc Nat R zero
term  acc0 = acc aux0
term  accR : (n : Nat) -> Acc Nat R n
{ accR (zero) = acc0
; accR (succ (zero)) = acc1
; accR (succ (succ n)) = acc2 n
}
term  acc_dest : .[n : Nat] -> (p : Acc Nat R n) -> (m : Nat) -> R m n -> Acc Nat R m
{ acc_dest [n] (acc p) = p
}
term  f : (x : Nat) -> Acc Nat R x -> Nat
{ f x (acc p) = case x : Nat
                { zero -> f (succ x) (p (succ x) r2)
                ; succ (zero) -> f (succ x) (p (succ x) (r1 [zero]))
                ; succ (succ y) -> zero
                }
}
term  h : (x : Nat) -> .[Acc Nat R x] -> Nat
{ h (zero) [acc [p]] = h (succ zero) [p (succ zero) r2]
; h (succ (zero)) [acc [p]] = h (succ (succ zero)) [p (succ (succ zero)) (r1 [zero])]
; h (succ (succ y)) [p] = zero
}
term  bla : Nat
term  bla = f zero acc0
type  Id : ^(A : Set) -> ^(a : A) -> ^ A -> Set
term  refl : .[A : Set] -> .[a : A] -> < refl : Id A a a >
error during typechecking:
p1
/// checkExpr 0 |- \ p -> refl : (p : Acc Nat R zero) -> Id Nat (h zero [p]) (h zero [acc0])
/// checkForced fromList [] |- \ p -> refl : (p : Acc Nat R zero) -> Id Nat (h zero [p]) (h zero [acc0])
/// new p : (Acc Nat R zero)
/// checkExpr 1 |- refl : Id Nat (h zero [p]) (h zero [acc0])
/// checkForced fromList [(p,0)] |- refl : Id Nat (h zero [p]) (h zero [acc0])
/// leqVal' (subtyping) [(p,0)] |- < refl : Id Nat (h zero [p]) (h zero [p]) >  <=+  Id Nat (h zero [p]) (h zero [acc0])
/// leqVal' (subtyping) [(p,0)] |- Id Nat (h zero [p]) (h zero [p])  <=+  Id Nat (h zero [p]) (h zero [acc0])
/// leqVal' [(p,0)] |- h zero [p]  <=^  zero : Nat
/// leqApp: (h {zero {p = v0}} {[p] {p = v0}})[] !<=^ zero{}[]
