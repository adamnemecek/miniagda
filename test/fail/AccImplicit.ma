data Acc ( A : Set) (Lt : A -> A -> Set) *(b : A) : Set
{ acc :  (accOut : (a : A) -> Lt a b -> Acc A Lt a) -> Acc A Lt b
}

data Nat : Set
{ zero : Nat
; succ : Nat -> Nat
}

{- R (S x) x  if x < 2
 -}
data R : Nat -> Nat -> Set
{ r1 : (x : Nat) -> R (succ (succ x)) (succ zero)
; r2 : R (succ zero) zero
}

{-
fun succR : (n : Nat) -> R (succ n) n
{ succR zero = r2
; succR (succ n) =
-}

let acc2 : (n : Nat) -> Acc Nat R (succ (succ n))
  = \ n -> acc (\ a -> \ p -> case p {})

fun aux1 : (a : Nat) -> (p : R a (succ zero)) -> Acc Nat R a
{ aux1 (succ (succ x)) (r1 .x) = acc2 x
}

let acc1 : Acc Nat R (succ zero)
  = acc aux1

fun aux0 : (a : Nat) -> (p : R a zero) -> Acc Nat R a
{ aux0 .(succ zero) r2 = acc1
}

let acc0 : Acc Nat R zero
  = acc aux0

fun accR : (n : Nat) -> Acc Nat R n
{ accR zero = acc0
; accR (succ zero) = acc1
; accR (succ (succ n)) = acc2 n
}

fun acc_dest : [n : Nat] -> (p : Acc Nat R n) ->
               (m : Nat) -> R m n -> Acc Nat R m
{ acc_dest n (acc p) = p
}

fun f : (x : Nat) -> Acc Nat R x -> Nat
{ f x (acc p) = case x
  { zero -> f (succ x) (p (succ x) r2)
  ; (succ zero) -> f (succ x) (p (succ x) (r1 zero))
  ; (succ (succ y)) -> zero
  }
}

{-
-- In Coq, g and h are accepted by the termination checker
fun g : (x : Nat) -> [Acc Nat R x] -> Nat
{ g x p = case x
  { zero -> g (succ x) (acc_dest zero p (succ x) r2)
  ; (succ zero) -> g (succ x) (acc_dest (succ zero) p (succ x) (r1 zero))
  ; (succ (succ y)) -> zero
  }
}

fun h : (x : Nat) -> [Acc Nat R x] -> Nat
{ h zero p = h (succ zero) (acc_dest zero p (succ zero) r2)
; h (succ zero) p = h (succ (succ zero)) (acc_dest (succ zero) p (succ (succ zero)) (r1 zero))
; h (succ (succ y)) p = zero
}
-}

fun h : (x : Nat) -> [Acc Nat R x] -> Nat
{ h zero (acc p) = h (succ zero) (p (succ zero) r2)
; h (succ zero) (acc p) = h (succ (succ zero)) (p (succ (succ zero)) (r1 zero))
; h (succ (succ y)) p = zero
}
{- The definition of h should be fine since

   q : Acc Nat R zero   iff  q = acc .Nat .R zero p

so the forced match does not refine the type [Acc Nat R x] further.
This means that h can be translated to case trees without any case on q,
it just uses the destructor. -}

eval let bla : Nat
  = f zero acc0

data Id (A : Set) (a : A) : A -> Set
{ refl : Id A a a
}

let p1 : (p : Acc Nat R zero) -> Id Nat (h zero p) (h zero acc0)
       = \ p -> refl
{- In a case tree representation of h this would type check! -}
