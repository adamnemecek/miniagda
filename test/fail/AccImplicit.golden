MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "AccImplicit.ma" ---
--- scope checking ---
--- type checking ---
type  Acc : ^(A : Set) -> ^(Lt : A -> A -> Set) -> (b : A) -> Set
term  Acc.acc : .[A : Set] -> .[Lt : A -> A -> Set] -> .[b : A] -> ^(accOut : (a : A) -> Lt a b -> Acc A Lt a) -> < Acc.acc accOut : Acc A Lt b >
term  accOut : .[A : Set] -> .[Lt : A -> A -> Set] -> (b : A) -> (acc : Acc A Lt b) -> (a : A) -> Lt a b -> Acc A Lt a
{ accOut [A] [Lt] b (Acc.acc #accOut) = #accOut
}
type  Nat : Set
term  Nat.zero : < Nat.zero : Nat >
term  Nat.succ : ^(y0 : Nat) -> < Nat.succ y0 : Nat >
type  R : ^ Nat -> ^ Nat -> Set
term  R.r1 : .[x : Nat] -> < R.r1 x : R (Nat.succ (Nat.succ x)) (Nat.succ Nat.zero) >
term  R.r2 : < R.r2 : R (Nat.succ Nat.zero) Nat.zero >
term  acc2 : (n : Nat) -> Acc Nat R (Nat.succ (Nat.succ n))
term  acc2 = \ n -> Acc.acc (\ a -> \ p -> case p : R a (Nat.succ (Nat.succ n))
                              {})
term  aux1 : (a : Nat) -> (p : R a (Nat.succ Nat.zero)) -> Acc Nat R a
{ aux1 (Nat.succ (Nat.succ x)) (R.r1 [.x]) = acc2 x
}
term  acc1 : Acc Nat R (Nat.succ Nat.zero)
term  acc1 = Acc.acc aux1
term  aux0 : (a : Nat) -> (p : R a Nat.zero) -> Acc Nat R a
{ aux0 .(succ zero) R.r2 = acc1
}
term  acc0 : Acc Nat R Nat.zero
term  acc0 = Acc.acc aux0
term  accR : (n : Nat) -> Acc Nat R n
{ accR Nat.zero = acc0
; accR (Nat.succ Nat.zero) = acc1
; accR (Nat.succ (Nat.succ n)) = acc2 n
}
term  acc_dest : .[n : Nat] -> (p : Acc Nat R n) -> (m : Nat) -> R m n -> Acc Nat R m
{ acc_dest [n] (Acc.acc p) = p
}
term  f : (x : Nat) -> Acc Nat R x -> Nat
{ f x (Acc.acc p) = case x : Nat
                    { Nat.zero -> f (Nat.succ x) (p (Nat.succ x) R.r2)
                    ; Nat.succ Nat.zero -> f (Nat.succ x) (p (Nat.succ x) (R.r1 [Nat.zero]))
                    ; Nat.succ (Nat.succ y) -> Nat.zero
                    }
}
term  h : (x : Nat) -> .[Acc Nat R x] -> Nat
{ h Nat.zero [Acc.acc [p]] = h (Nat.succ Nat.zero) [p (Nat.succ Nat.zero) R.r2]
; h (Nat.succ Nat.zero) [Acc.acc [p]] = h (Nat.succ (Nat.succ Nat.zero)) [p (Nat.succ (Nat.succ Nat.zero)) (R.r1 [Nat.zero])]
; h (Nat.succ (Nat.succ y)) [p] = Nat.zero
}
term  bla : Nat
term  bla = f Nat.zero acc0
type  Id : ^(A : Set) -> ^(a : A) -> ^ A -> Set
term  Id.refl : .[A : Set] -> .[a : A] -> < Id.refl : Id A a a >
error during typechecking:
p1
/// checkExpr 0 |- \ p -> refl : (p : Acc Nat R Nat.zero) -> Id Nat (h Nat.zero [p]) (h Nat.zero [acc0])
/// checkForced fromList [] |- \ p -> refl : (p : Acc Nat R Nat.zero) -> Id Nat (h Nat.zero [p]) (h Nat.zero [acc0])
/// new p : (Acc Nat R Nat.zero)
/// checkExpr 1 |- refl : Id Nat (h Nat.zero [p]) (h Nat.zero [acc0])
/// checkForced fromList [(p,0)] |- refl : Id Nat (h Nat.zero [p]) (h Nat.zero [acc0])
/// leqVal' (subtyping)  < Id.refl : Id Nat (h Nat.zero [p]) (h Nat.zero [p]) >  <=+  Id Nat (h Nat.zero [p]) (h Nat.zero [acc0])
/// leqVal' (subtyping)  Id Nat (h Nat.zero [p]) (h Nat.zero [p])  <=+  Id Nat (h Nat.zero [p]) (h Nat.zero [acc0])
/// leqVal'  h Nat.zero p  <=^  Nat.zero : Nat
/// leqApp: head mismatch h != Nat.zero
